### PR TITLE
[stable/jenkins] Add support for custom pod templates

### PR DESCRIPTION
**Pull request by lgg42**
Signed-off-by: Luis Garnica Guilarte <luisgarnica42@gmail.com>

#### Is this a new chart
No.
> NOTE: We're experiencing a high volume of PRs to this repo and reviews will be delayed. Please host your own chart repository and submit your repository to the Helm Hub instead of this repo to make them discoverable to the community. [Here](https://github.com/helm/hub/blob/master/Repositories.md) is how to submit new chart repositories to the Helm Hub.

#### What this PR does / why we need it:
This PR allows to configure additional pod templates in the default kubernetes cloud.
We want to configure from scratch custom pod templates the same way we do with all the settings using JCasC.

#### Which issue this PR fixes
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
  - fixes #

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] [DCO](https://github.com/helm/charts/blob/master/CONTRIBUTING.md#sign-your-work) signed
- [x] Chart Version bumped
- [x] Variables are documented in the README.md
- [x] Title of the PR starts with chart name (e.g. `[stable/mychartname]`)

### CODE REVIEW

File: stable/jenkins/README.md [hunk 1]
```diff
@@ -189,6 +189,7 @@ The following tables list the configurable parameters of the Jenkins chart and t
 | `serviceAccountAgent.name`        | name of the agent ServiceAccount to be used by access-controlled resources | autogenerated |
 | `serviceAccountAgent.create`      | Configures if an agent ServiceAccount with this name should be created | `false`         |
 | `serviceAccountAgent.annotations` | Configures annotation for the agent ServiceAccount | `{}`                              |
+| `podTemplates`                    | Configures extra pod templates for the default kubernetes cloud | `{}`                              |
```
Comment (wmcdona89) [inline stable/jenkins/README.md]: What about putting this under `master.JCasC.podTemplates` so all the JCasC config values would be grouped together. Another option could be to put it in the agent section as `podTemplates`.

Comment (lgg42) [reply stable/jenkins/README.md]: I also think is a nice change. I left a comment in your PR. I think maybe your PR is the one to go. Let's see how we manage to get this feature working the best possible way. I'm on hold (on committing) until we find common ground :smile:

Comment (lgg42) [reply stable/jenkins/README.md]: Change done to `agent.podTemplates` as we talked in your PR.

File: stable/jenkins/templates/_helpers.tpl [hunk 2]
```diff
@@ -142,6 +142,11 @@ jenkins:
         yaml: |-
           {{ tpl .Values.agent.yamlTemplate . | nindent 10 | trim }}
         yamlMergeStrategy: "override"
+      {{- if .Values.agent.podTemplates }}
+        {{- range $key, $val := .Values.agent.podTemplates }}
+          {{ tpl $val $| nindent 6 }}
```
Comment (wmcdona89) [inline stable/jenkins/templates/_helpers.tpl]: how about `{{-` to avoid the blank line and `$ |` with a space

File: stable/jenkins/CHANGELOG.md [hunk 3]
```diff
@@ -5,6 +5,10 @@ numbering uses [semantic versioning](http://semver.org).
 
 NOTE: The change log until version 1.5.7 is auto generated based on git commits. Those include a reference to the git commit to be able to get more details.
 
+## 1.10.2 Add support for configuring custom pod templates
+
+Add `agent.podTemplates` option for declaring custom pod templates in the default configured kubenetes cloud.
```
Comment (wmcdona89) [inline stable/jenkins/CHANGELOG.md]: nit: "kubernetes"

File: stable/jenkins/README.md [hunk 4]
```diff
@@ -482,6 +483,40 @@ master:
 Docs taken from https://github.com/jenkinsci/docker/blob/master/Dockerfile:
 _Jenkins is run with user `jenkins`, uid = 1000. If you bind mount a volume from the host or a data container,ensure you use the same uid_
 
+## Adding custom pod templates
+It is possible to add custom pod templates for the default configured kubernetes cloud.
+Add a key under `agent.podTemplates` for each pod template. Each key (prior to | character) is just a label, and can be any value.
+Keys are only used to give the pod template a meaningful name.  The only restriction is they may only contain RFC 1123 \ DNS label
+characters: lowercase letters, numbers, and hyphens. Each pod template can contain multiple containers.
+There's no need to add the *jlnp* container since the kubernetes plugin will automatically inject it into the pod.
```
Comment (wmcdona89) [inline stable/jenkins/README.md]: nit: jnlp

File: stable/jenkins/values.yaml [hunk 5]
```diff
@@ -470,6 +470,33 @@ agent:
...
+  #          command: "/bin/sh -c"
+  #          args: "cat"
+  #          ttyEnabled: true
+  #          privileged: true
+  #          resourceRequestCpu: "400m"
+  #          resourceRequestMemory: "512Mi"
+  #          resourceLimitCpu: "1"
+  #          resourceLimitMemory: "1024Mi"
+
+
```
Comment (wmcdona89) [inline stable/jenkins/values.yaml]: nit: probably don't need 2 blank lines here

**Review by wmcdona89 — CHANGES_REQUESTED** [review]

**Comment by wmcdona89** [comment]
@lgg42 Turns out I'll be able to leverage this to inject "agents" into the default casc. Here's an example with the new podTemplate template in my PR.

```
# A list of agents. Each entry corresponds to a chart `agent` in terms of the values that can be set.
additionalAgents:
  - podName: maven
    customJenkinsLabels: maven
    image: jenkins/jnlp-agent-maven
  - podName: node
    customJenkinsLabels: node
    image: jenkins/jnlp-agent-node

agent:
  podTemplates:
    agents: |-
      {{- /* save chart agent as `set` modifies the input dictionary */}}
      {{- $rootAgent := .Values.agent }}

      {{- /* each additional agent inherits from $rootAgent */}}
      {{- range .Values.additionalAgents }}
      {{- include "jenkins.casc.podTemplate" (set $ "Values" (set $.Values "agent" (merge . $rootAgent))) }}
      {{- end }}

      {{- /* restore chart agent */}}
      {{- $_ := set .Values "agent" $agent }}
```

File: stable/jenkins/values.yaml (hunk 5 above)
Comment (lgg42) [reply stable/jenkins/values.yaml]: thanks! corrected!

File: stable/jenkins/README.md (hunk 4 above)
Comment (lgg42) [reply stable/jenkins/README.md]: thanks! corrected!

File: stable/jenkins/CHANGELOG.md (hunk 3 above)
Comment (lgg42) [reply stable/jenkins/CHANGELOG.md]: done! very keen eye!

File: stable/jenkins/templates/_helpers.tpl (hunk 2 above)
Comment (lgg42) [reply stable/jenkins/templates/_helpers.tpl]: Tomorrow I'll change/test this! thank you so much for your recommendations!

**Comment by lgg42** [comment]
> @lgg42 Turns out I'll be able to leverage `agent.podTemplates` to inject "agents" into the default casc. Here's an example with the new podTemplate template in my PR.
> 
> ```
> # A list of agents. Each entry corresponds to a chart `agent` in terms of the values that can be set.
> additionalAgents:
>   - podName: maven
>     customJenkinsLabels: maven
>     image: jenkins/jnlp-agent-maven
>   - podName: node
>     customJenkinsLabels: node
>     image: jenkins/jnlp-agent-node
> 
> agent:
>   podTemplates:
>     agents: |-
>       {{- /* save chart agent as `set` modifies the input dictionary */}}
>       {{- $rootAgent := .Values.agent }}
> 
>       {{- /* each additional agent inherits from $rootAgent */}}
>       {{- range .Values.additionalAgents }}
>       {{- include "jenkins.casc.podTemplate" (set $ "Values" (set $.Values "agent" (merge . $rootAgent))) }}
>       {{- end }}
> 
>       {{- /* restore chart agent */}}
>       {{- $_ := set .Values "agent" $agent }}
> ```

Amazing!!! :muscle:

**Comment by arencibiafrancisco** [comment]
Can't wait for this to be merged! Have been looking for this! Another JenkinsX user here. `agent.podTemplates` is perfect!

File: stable/jenkins/templates/_helpers.tpl (hunk 2 above)
Comment (lgg42) [reply stable/jenkins/templates/_helpers.tpl]: Thank you also for this correction! done already.

File: stable/jenkins/values.yaml [hunk 6]
```diff
@@ -470,6 +470,31 @@ agent:
...
+  # Add a key under podTemplates for each pod template. Each key (prior to | character) is just a label, and can be any value.
+  # Keys are only used to give the pod template a meaningful name.  The only restriction is they may only contain RFC 1123 \ DNS label
+  # characters: lowercase letters, numbers, and hyphens. Each pod template can contain multiple containers.
+  # For this pod templates configuration to be loaded the following values must be set:
+  # master.JCasC.enabled: true
+  # master.JCasC.defaultConfig: true
+  # master.JCasC.configScripts: at least one key (cannot be empty {})
+  # Best reference is https://<jenkins_url>/configuration-as-code/reference#Cloud-kubernetes. The example below creates a python pod template.
+  podTemplates: {}
+  #  python: |
```
Comment (wmcdona89) [inline stable/jenkins/values.yaml:483]: I just noticed the JenkinsX [values](https://github.com/jenkins-x/jenkins-x-platform/blob/731a37428e057501e294d2737ebff785052e1fcf/jenkins-x-platform/values.yaml#L1057) doesn't make use of multi-line strings. If it's not truly needed then I'd vote to just make a toYaml call in the template.

File: stable/jenkins/README.md [hunk 7]
```diff
@@ -482,6 +483,40 @@ master:
...
+```
+master.JCasC.enabled: true
+master.JCasC.defaultConfig: true
+master.JCasC.configScripts: at least one key (cannot be empty {})
+```
+The example below creates a python pod template in the kubernetes cloud.
+```
+agent:
+  podTemplates:
+    python: |
```
Comment (wmcdona89) [inline stable/jenkins/README.md:501]: FYI. I noticed the [JenkinsX values file](https://github.com/jenkins-x/jenkins-x-platform/blob/731a37428e057501e294d2737ebff785052e1fcf/jenkins-x-platform/values.yaml#L1057) doesn't use multi-line strings. As it turns out, having this be a multi-line string (and not calling `toYaml` in the template) actually provides additional flexibility and allows for passing in a block of templating.

```
agent:
  podTemplates:
    agents: |-
      {{- /* save chart agent as `set` modifies the input dictionary */}}
      {{- $rootAgent := .Values.agent }}

      {{- /* each additional agent inherits from $rootAgent */}}
      {{- range .Values.additionalAgents }}
      {{- include "jenkins.casc.podTemplate" (set $ "Values" (set $.Values "agent" (merge . $rootAgent))) }}
      {{- end }}

      {{- /* restore chart agent */}}
      {{- $_ := set .Values "agent" $rootAgent }}
```

Comment (lgg42) [reply stable/jenkins/README.md:501]: I'm glad you pointed me this out and resulted to be a good thing. I just followed examples from other parts of this chart.
So, does it looks already ok to test to you? 
It's my first contribution here, how long does this process tend to take?

File: stable/jenkins/README.md [hunk 8]
```diff
@@ -482,6 +483,40 @@ master:
...
+It is possible to add custom pod templates for the default configured kubernetes cloud.
+Add a key under `agent.podTemplates` for each pod template. Each key (prior to | character) is just a label, and can be any value.
+Keys are only used to give the pod template a meaningful name.  The only restriction is they may only contain RFC 1123 \ DNS label
+characters: lowercase letters, numbers, and hyphens. Each pod template can contain multiple containers.
+There's no need to add the *jnlp* container since the kubernetes plugin will automatically inject it into the pod.
+For this pod templates configuration to be loaded the following values must be set:
+```
+master.JCasC.enabled: true
+master.JCasC.defaultConfig: true
+master.JCasC.configScripts: at least one key (cannot be empty {})
```
Comment (torstenwalter) [inline stable/jenkins/README.md]: This line can be removed

File: stable/jenkins/values.yaml [hunk 9]
```diff
@@ -470,6 +470,31 @@ agent:
...
   slaveConnectTimeout: 100
 
+  # Below is the implementation of custom pod templates for the default configured kubernetes cloud.
+  # Add a key under podTemplates for each pod template. Each key (prior to | character) is just a label, and can be any value.
+  # Keys are only used to give the pod template a meaningful name.  The only restriction is they may only contain RFC 1123 \ DNS label
+  # characters: lowercase letters, numbers, and hyphens. Each pod template can contain multiple containers.
+  # For this pod templates configuration to be loaded the following values must be set:
+  # master.JCasC.enabled: true
+  # master.JCasC.defaultConfig: true
+  # master.JCasC.configScripts: at least one key (cannot be empty {})
```
Comment (torstenwalter) [inline stable/jenkins/values.yaml]: This line can be removed.

File: stable/jenkins/Chart.yaml [hunk 10]
```diff
@@ -1,7 +1,7 @@
 apiVersion: v1
 name: jenkins
 home: https://jenkins.io/
-version: 1.10.1
+version: 1.10.2
```
Comment (torstenwalter) [inline stable/jenkins/Chart.yaml]: This added functionality is worth a minor version bump ;-)

**Review by torstenwalter — CHANGES_REQUESTED** [review]
This looks good to me. I have just some minor remarks - see inline comments.
Could you do me a favor and add `defaultConfig: true` in `ci/casc-values.yaml`. I think it"s time to check these setting as part of the build.

File: stable/jenkins/Chart.yaml (hunk 10 above)
Comment (lgg42) [reply stable/jenkins/Chart.yaml]: Oh yeah, so true! done!

File: stable/jenkins/values.yaml (hunk 9 above)
Comment (lgg42) [reply stable/jenkins/values.yaml]: Removed.

File: stable/jenkins/README.md (hunk 8 above)
Comment (lgg42) [reply stable/jenkins/README.md]: Removed.

**Comment by lgg42** [comment]
@torstenwalter thanks for your change comments! all resolved. I also added the line you requested in `ci/casc-values.yaml`.

**Comment by lgg42** [comment]
CI's failing with 
```
I0404 04:49:59.725] Error: incompatible versions client[v2.16.5] server[v2.15.2]
```
Is this something I could fix from my part?

**Comment by torstenwalter** [comment]
@lgg42 Thanks a lot for the PR.

**Comment by lgg42** [comment]
Thank you for allowing this PR go through! ♥️

**Pull request merged**